lib/zk_recipes/cache.rb

```diff
@@ -205,31 +508,148 @@ def process_pending_updates
         return
       end
 
-      @logger&.debug { "processing pending updates=#{@pending_updates.size}" }
-      @pending_updates.reject! do |missed_path, _|
-        update_cache(missed_path)
+      @logger&.debug { "process_pending_updates: updates=#{@pending_updates.size}" }
+      @pending_updates.each do |path, type|
+        @logger&.debug { "process_pending_updates: path=#{path} type=#{type}" }
+        delete = case type
+                 when :static
+                   update_static_path(path)
+                 when :runtime
+                   update_runtime_path(path)
+                 when :directory
+                   update_directory_path(path)
+                 else
+                   @logger&.error { "process_pending_updates: unknown pending_update type=#{type.inspect}" }
+                   true
+                 end
+        @pending_updates.delete(path) if delete
+      end
+
+      if @pending_updates.empty?
+        @logger&.debug("process_pending_updates: processed all pending updates")
+      else
+        @logger&.warn { "process_pending_updates: failed to process pending updates=#{@pending_updates.inspect}" }
       end
     end
 
-    class CachedPath
-      attr_reader :value, :stat
-      def initialize(value, stat: nil, valid: false)
-        @value = value
+    class StaticPath
+      def initialize(default_value, deserializer)
+        @mutex = Mutex.new
+        @value = @default_value = default_value
+        @deserializer = deserializer
+        @stat = nil
+        @valid = false
+      end
+
+      def value
+        @mutex.synchronize { @value }
```

Comment:
Does this need to be in a mutex? I think a simple read should be safe outside the mutex. The worst that can happen is reading a stale value, yes?